pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -430,6 +438,165 @@ func consumerRefMatches(consumer *corev1.ObjectReference, machine *machinev1.Mac
+	}
+	// Fetch IgnitionURL and the CA cert
+	ignitionURL := ignData.Ignition.Config.Append[0].Source
+	caCertRaw := ignData.Ignition.Security.TLS.CertificateAuthorities[0].Source
+	caCertB64 := strings.TrimPrefix(caCertRaw, "data:text/plain;charset=utf-8;base64,")
+
+	// We didn't get any URL
+	if ignitionURL == "" {
+		err := fmt.Errorf("error extracting full Ignition URL")
+		log.Printf("URL not found : %s", err.Error())
+		return err
+	}
+	// We didn't get the URL's Cert
+	if caCertB64 == "" {
+		err := fmt.Errorf("CA cert for full Ignition URL was not specified")
+		log.Printf("Cert not found in ignition data: %s", err.Error())
+		return err
+	}
+	caCert, err := base64.StdEncoding.DecodeString(caCertB64)
+	if err != nil {
+		log.Printf("Could not base64 decode caCert: %s", err.Error())
+		return fmt.Errorf("could not base64 decode caCert: %w", err)
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+	client := retryablehttp.NewClient()
+	client.HTTPClient.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: caCertPool,
+		},
+	}
+
+	// Get the full ignition from the URL
+	resp, err := client.Get(ignitionURL)
+	if err != nil {
+		log.Printf("Error fetching full ignition from URL : %s", err.Error())
+		return fmt.Errorf("error fetching full ignition from URL: %w", err)
+	}
+	defer resp.Body.Close()
+
+	var fullIgnition []byte
+	fullIgnition, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Printf("Error reading full ignition data from URL : %s", err.Error())
+		return fmt.Errorf("error reading full ignition data from URL: %w", err)
+	}
+
+	// Create a new secret with ignition data
+	machineUserDataSecret = createNewMachineSecret(fullIgnition, machine)
+	err = a.client.Create(ctx, machineUserDataSecret)
+	if err != nil {
+		log.Printf("Cannot create the %s secret: %s", machine.Name+"-user-data", err.Error())
+		return fmt.Errorf("cannot create the %s secret: %w", machine.Name+"user-data", err)
+	}
+
+	// Set Machine's secretreference to the newly created secret
+	config.UserData = &corev1.SecretReference{
+		Name:      machine.Name + "-user-data",
+		Namespace: machine.Namespace,
+	}
+
+	return nil
+}
+
+// createNewSecret uses Full Ignition data and creates a '-user-data' suffixed secret per machine
+func createNewMachineSecret(data []byte, machine *machinev1.Machine) *corev1.Secret {
+	// Create a new Secret with Full Ignition
+	return &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      machine.Name + "-user-data",
+			Namespace: machine.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				metav1.OwnerReference{
+					Controller: pointer.BoolPtr(true),
+					Kind:       "Machine",
+					Name:       machine.Name,
+					APIVersion: machine.APIVersion,
+					UID:        machine.UID,
+				},
+			},
+			Finalizers: []string{machinev1.MachineFinalizer},
```

Comment:
Why does this need a finalizer? And what removes it?

Reply:
Because we want to remove this secret when the host disappears.

Reply:
When a host disappears, the Machine will also get deleted, and the owner ref would cause this to get garbage-collected, right?

If you do keep this finalizer, something has to remove it after it is marked for deletion. What does that?

Reply:
Its probably me overthinking the garbage collection. I was trying to use the same logic used here: https://github.com/openshift/cluster-api-provider-baremetal/blob/bc712c5ef889156e665fe19a31736209a24567cb/pkg/cloud/baremetal/actuators/machine/actuator.go#L126
Its probably not needed since we already have the ownerreference set. So, will remove it.